### ISSUE TITLE
[FEATURE REQUEST] - Script usage for the average user

### ISSUE THREAD

**Issue by Nick-Mazuk** [finale-lua/lua-scripts#17]
So @CJGarciaMusic messaged me about a very valid point with the Library and how it's currently set up.

**And discussing potential solutions soon is a good idea so we won't need to refactor thousands of lines of code later on.**

Essentially, it's hard for users to use any of the main scripts. They either need to:

- Use `git` to clone the repo and update scripts
- They couldn't necessarily download a single script (if it uses the Library)
    - They would also need to download the Library, make sure it's in a subdirectory, and update both files whenever there's an update.

@CJGarciaMusic, correct me if I'm mistaken about anything you were saying.

This is a valid point that dramatically diminishes the impact of these scripts as things stand because many people won't want to go through that process or don't know `git` to get started.

Now, I'd rather not completely get rid of the Library as it stands because it can be incredibly helpful with development. So here are possibilities I thought of to resolve this issue. Thoughts and ideas appreciated.

---

1. **Have a "Dropbox Installer"**. Ok, for those who don't know how / don't want to use git, I can set up a Dropbox folder we can share with people. That folder can mainly just be the repo. Then I can set up a cron job with Keyboard Maestro on my computer to periodically git pull and update the Dropbox folder. Then people can get automatic updates. I've used this with several of my personal scripts with several people, and it's gone decently well. We'd need some proper documentation with the JW Lua plug-in installation. But people would need that anyways. The only challenge here would be notifying people of new scripts. Previously installed scripts will be updated automatically. This is by far the easiest option I've thought of and will only take ~30 minutes of work on my end.
2. **We create a JW Lua Script as an installer/updater**. Assuming JW Lua can connect to the internet or at least use curl, this should be possible. I have a rough outline of the code in my head. Then, people can set up a single JW Lua script, find new scripts, install the new scripts, and update them from a single script. I'm not sure if JW Lua can connect to the internet. With regular Lua, you need to import a module. Not sure if JW Lua includes that module. Of all the options I've thought of, this would likely be the hardest and most time-consuming. But probably the best user experience.
3. **Create a website and pre-compile the scripts**. Lately, I've been doing a surprising amount of web development, so this one is definitely doable. Essentially, you would be able to explore scripts on the site. When you download a script, the site will then look for all the imports and just replace the import statement with the contents of the library module. Crude, but it will work. Updates are going to be harder with this method, though. If we choose this route, I'd rather code the site. Which means there won't be a GUI to update it. So anyone updating the site would need to know HTML, CSS, and JS. The timeline would probably be ~3-4 weeks for the first version. Realistically, it would be only about 10 hours total, but don't want to overpromise a delivery date.
4. **Perhaps @rpatters1 could create an installer plug-in**? If JW Lua can't access the internet, I'm sure a plug-in written in C++ could. It would put most of the work on his shoulders since I don't know C++, but it's a potential option. @rpatters1, don't feel obligated to do this. Just throwing out ideas.
5. **Alternatively, forget about good user experience**. Kind of defeats the purpose of a public repo, but it's an option. I'd rather not choose this option, but it's an option.

---

I think, ideally, it would be great to have the best possible developer experience and user experience.

I don't think it's worth abandoning the idea of a Library where we can import functions at this point. If we can't find a suitable way for non-techie users to use the scripts, then that's probably time. I don't think we're there yet.

Thoughts? @CJGarciaMusic, @rpatters1

**Comment by rpatters1** [comment]
I share @CJGarciaMusic's concern.

Another option would be to use git. Perhaps we could package up the appropriate git commands to clone (if needed) and pull master in a simple shell script. Then all the potential user would have to do it execute the shell script.

**Comment by rpatters1** [comment]
To follow up: GitHub is almost there. If you click on the Clone->Zip file option, you get the master branch in a git-enabled folder that you can put anywhere you like.

**Comment by rpatters1** [comment]
Another idea: JW Lua may not be able to access the internet, but perhaps it can execute the shell script. Then the installer could be a lua scrupt.

**Comment by rpatters1** [comment]
After a little research I realize that, unlike Mac and Linux, git does not come preinstalled on Win. But this link works for anyone I believe:

https://github.com/Nick-Mazuk/jw-lua-scripts/archive/master.zip

So there could be a link on a web page that downloaded that link.

**Comment by Nick-Mazuk** [comment]
Looks like Lua **can** run shell commands. For instance, I just ran:

```lua
os.execute('cd /Users/nickmazuk/Desktop && mkdir lua_test && cd lua_test && curl https://google.com --output google.html')
```

and it successfully created a folder on my Desktop and saved the Google Hompage into that folder!

That's neat.

Given this, I'd say probably the best solution is to use Lua to create an installer and updater. Thoughts?

**Comment by rpatters1** [comment]
I like it. I think there is a PDK Framework command to open a dialog to get the target folder. I don't suppose there is some way for a script to add to the JW Lua menu in Finale.

**Comment by rpatters1** [comment]
Also, we'll need to test it on Win as well as Mac. I can do that easily. But for reference, I do that on my Mac using VMWare, in case y'all ever wanted to test on Win for yourselves.

**Comment by Nick-Mazuk** [comment]
I've tried using a VM to use Windows before. It was very buggy with JW Lua specifically, though. That may just be my experience, though.

There is a JW Lua command to have the user select a folder to upload files into (`FCFileSaveAsDialog`). I've used it before. I've also saved config options to a TXT file. So perhaps we could use that to save the file path (that way users don't need to mess around with preferences).

Alternatively, we could set the default save path to be their `Plug-ins/JW Lua/a folder name we decide on/`. There's a JW Lua function to find their plugin folder (`FCString():SetPluginsFolderPath()`).

I don't know how to add a script to the JW Lua menu through a JW Lua script. That would be incredibly useful! I know the perfect layout plugin packages a copy of JW Lua with the Perfect Layout plugin already in the menu bar. So perhaps there's a way to modify the `.bundle` file to add the JW Lua script?

Also, in case Windows need a different shells script than macOS, we can use `FCUI:IsOnMac()` and `FCUI:IsOnWindows()` to distinguish between the two.

**Comment by rpatters1** [comment]
I haven't used JW Lua on Windows enough to have an opinion about how buggy
it is. (Honestly, I would be surprised if it's any more buggy than on
native Win. But it is definitely slower.) I was able to test the library
loading without any problems.

I mean, ultimately if this project is really going to take off, we have to
get Jari on board, at least to the extent of allowing me to work on his
code if he is not able to. We could address a tremendous amount of
headaches with some simple changes to JW Lua. We could eliminate rafts of
lua code with simple enhancements to the PDK Framework. And if we are
unable to build JW Lua soon with the latest PDK, Finale will gradually
begin to leave it behind.

All by way of saying, all of us need to keep an eagle eye out on Facebook
for Jari. If he posts, try to get his attention. I saw a post from him a
few weeks ago in the Scripts group I believe. I posted a greeting, but
unfortunately he did not reply.

**Comment by Nick-Mazuk** [comment]
I agree and that sounds like a plan.

So here are some ideas for next steps:

1. Get other JW Lua creators on board, that way we can really build up a library of scripts and get more opinions from other developers
    - I'll email Jan, Harold, and the French coder (don't know his name) later today and post in the JW Lua email group
2. Figure out the exact functions for the installer
    - Are we going to download the entire repo?
    - Download scripts individually?
    - etc.
3. Create a JW Lua script for what users are going to see. Essentially, figure out the UI. That way we know what data we need to get using a `curl` command or similar.
4. Figure out a good backend to fetch the info needed for the installer/updater. Potentially it could simply be downloading the repo and we won't need to do anything special. But it depends on the exact functionality we decide on.
5. Hook everything up and test it!

Thoughts?

**Comment by rpatters1** [comment]
To address your specific points, I don't think we should put lua scripts in
the plugins folder. On Mac, at least, plugins are in a privileged location
that is already restricted in Catalina so that you have to use root to copy
files there. (For some users.) My guess is that future macOS versions could
make it even harder to add ad hoc files to the plugins folder.

JW Lua is putting its data somewhere. I don't think it is inside the
bundle, but so far I'm not finding it. Will keep looking.

**Comment by CJGarciaMusic** [comment]
Here's a little proof of concept you might all enjoy... All it does is list and open the link. Downloading and installing shouldn't be too difficult...

`function plugindef()
    finaleplugin.RequireSelection = false
    finaleplugin.Author = "CJ Garcia"
    finaleplugin.Copyright = "© 2020 CJ Garcia Music"
    finaleplugin.Version = "0.1"
    finaleplugin.Date = "6/15/2020"
    finaleplugin.CategoryTags = "Note"
    return "Lua Repository", "Lua Repository", "A list of all current public Lua scripts."
end

local tmp_dir = "/tmp/"

if os.execute("cd "..tmp_dir.."lua_test.html") == true then
    os.execute("cd "..tmp_dir.."lua_test && curl https://github.com/Nick-Mazuk/jw-lua-scripts --output lua_list.html")
else
    os.execute("cd "..tmp_dir.." && curl https://github.com/Nick-Mazuk/jw-lua-scripts --output lua_list.html")
end

function titlecase(str)
    local buf = {}
    local in_word = false
    for i = 1, #str do
        local c = string.sub(str, i, i)
        if in_word then
            table.insert(buf, string.lower(c))
            if (string.find(c, '%s')) or (string.find(c, '%.')) then
                in_word = false
            end
        else
            table.insert(buf, string.upper(c))
            if (string.find(c, '%('))  then
                in_word = false
            else
                in_word = true
            end
        end
    end
    return table.concat(buf)
end

local current_scripts = {}
local download_links = {}
local current_file = tmp_dir.."/lua_list.html"

for line in io.lines (current_file) do
    if string.find(line, "title=\".*%.lua\"") and (string.find(line, ".lua\" id=\""))then
        local first_index = string.find(line, "title=\".*%.lua\"")
        local last_index = string.find(line, ".lua\" id=\"")
        local file_name = (string.sub(line, first_index + 7, last_index - 1))
        local display_name = titlecase(string.gsub(file_name, "_", " "))
        local download_link = "https://github.com/Nick-Mazuk/jw-lua-scripts/blob/master/"..file_name..".lua"
        table.insert(current_scripts, display_name)
        table.insert(download_links, download_link)
    end
end 

local str = finale.FCString()
 
str.LuaString = "Download Lua Script"
local dialog = finale.FCCustomWindow()
dialog:SetTitle(str)

local script_list = dialog:CreateListBox(0, 0)
script_list:SetWidth(244)


for key, value in pairs(current_scripts) do
    str.LuaString = value
    script_list:AddString(str)
end

dialog:CreateOkButton()
dialog:CreateCancelButton()
 
if dialog:ExecuteModal(nil) == 1 then
    local link_index = script_list:GetSelectedItem() + 1
    os.execute("open "..download_links[link_index])
end`

**Comment by rpatters1** [comment]
I found the location where JW Lua stores the menu items on Mac. It's
probably a parallel location on Windows.

~/Library/Application Support/Makemusic/Finale 26/jwpluginstorage.xml

So in theory you could add to that file programmatically. I'm not sure how
JW Lua would react to it while running, though.

**Comment by Nick-Mazuk** [comment]
Wow, you two are great!

I'll definitely have to try out @CJGarciaMusic's script.

I can confirm that on my machine, the JW Lua scripts are also stored in that location.

**Comment by rpatters1** [comment]
Another thought I had is that it's too bad we can't create subfolders that
reflect into submenus in the Finale plugin menu. But without access to JW
Lua source I don't see how it will be possible.

**Comment by Nick-Mazuk** [comment]
Yeah, it would be great if we could get submenus.

---

Anyways, I figured out something useful with the library. In the `jwpluginstorage.xml` file, there's this element:

``` xml
<UserPreferences SelectedPage="1" RunMultiSelectionMode="0" ExpCleanConsole="1" DevCleanConsole="1" Windowshade="0" EditorTabWidth="4" EditorFontName="Verdana" EditorFontSize="12" EditorTheme="1"></UserPreferences>
```

It's right after the `<PluginNodes>` element.

Inside, we can put required paths. In other words, we can add in the path to our library! So for instance, on my machine, it would be:

``` xml
<UserPreferences SelectedPage="1" RunMultiSelectionMode="0" ExpCleanConsole="1" DevCleanConsole="1" Windowshade="0" EditorTabWidth="4" EditorFontName="Verdana" EditorFontSize="12" EditorTheme="1">
    <RequireFolders CollectionIdx="1" FCString="/Users/nickmazuk/Documents/Github/jw-lua-scripts/library"/>
</UserPreferences>
```

_Note: If there are multiple required folders, they must have unique `CollectionIdx` values._

Then, instead of writing out this to import a library:

``` lua
local path = finale.FCString()
path:SetRunningLuaFolderPath()
package.path = package.path .. ";" .. path.LuaString .. "/library/?.lua"
local articulation = require("articulation")
```

…we can simply do

``` lua
local articulation = require("articulation")
```

So in short, when we ask for a folder to save the JW Lua files, we can also add the library path to this preference file.

**Comment by rpatters1** [comment]
So one of the simple changes I would like to see in JW Lua is that it
automatically adds the scripts folder to the list of running folders. As
long as we don't have access to changing the code, this seems like a
reasonable substitute. Therefore I would advocate adding our main folder to
the xml file and leaving the "library/" qualifications in the scripts. Not
only does this set us up for a hoped-for revision in JW Lua, it makes clear
that they are library scripts in the library folder.

**Comment by rpatters1** [comment]
So here is my proposal for how to add the libraries. We could refactor the library loads like this:

```
local path = finale.FCString()
path:SetRunningLuaFolderPath()
package.path = package.path .. ";" .. path.LuaString .. "?.lua"
local library = require("library.general_library")
```

This makes it clear that the libraries are in the ```library``` subfolder.  Then once we have a mechanism in place to add our local folder of ```jw-lua-script/?.lua``` to the ```jwpluginstorage.xml``` we can remove the first 3 lines leaving only

```local library = require("library.general_library")```

The beauty of this is that ultimately I would like for JW Lua itself to run scripts in an environment where the running script folder (plus ```?.lua```) is part of the runtime environment by default. Then we will not even need to modify ```jwpluginstorage.xml```.

**Comment by rpatters1** [comment]
I just did a proof of concept of my idea. That was an amazing find in ```jwpluginstorage.xml``` btw. I would never have figured it out from mine, because my ```UserPreferences``` branch contained no sub-branches. Anyway, I added it as you did:

```
<UserPreferences SelectedPage="2" RunMultiSelectionMode="0" ExpCleanConsole="1" DevCleanConsole="1" Windowshade="0" EditorTabWidth="4" EditorFontName="Verdana" EditorFontSize="12" EditorTheme="1">
       <RequireFolders CollectionIdx="1" FCString="<my-folder-location>/Lua Scripts"/>
</UserPreferences>
```

Now I can add the library with one line of code:

```local library = require("library.general_library")```

It seems lua treats "." as a path delimiter. This also works:

```local library = require("library/general_library")```

BTW, how do you get your code snippets to be color-labeled?

**Comment by Nick-Mazuk** [comment]
Great!

I think that's going to be the way to go. It's simple, and I do like how adding `require("library.general_library"` signals that the file is in the library folder (as opposed to some random Lua library). We just have to figure out how to add that code to the `UserPreferences`.

---

For the color-labeling, you just need to add the name of the language after ``` in the multiline code snippet.

For instance, this …

    ``` lua
    -- this is lua
    print('hello world')
    ```

… becomes …

``` lua
-- this is lua
print('hello world')
```

See **Syntax Highlighting** in the markdown cheatsheet: https://guides.github.com/features/mastering-markdown/

**Comment by rpatters1** [comment]
So the Windows location is exactly where I expected to find it. The analogous place on Windows (which I found by looking in Finale Preferences->Folders):

```
C:\Users\Robert Patterson\AppData\Roaming\MakeMusic\Finale 26\jwpluginstorage.xml
```
I then edited this file as we have done for the Mac. Everything worked as expected. I think this approach is the right one.

**Comment by rpatters1** [comment]
Something like this might allow us to edit the ```jwpluginstorage.xml``` either as a shell-out or directly in JW Lua:

[xml2lua](https://github.com/manoelcampos/xml2lua/blob/master/README.adoc)

**Comment by Nick-Mazuk** [comment]
I have an idea on how to add the file path to `jwpluginstorage.xml` without using a 3rd-party library. Will code it up later today.

Also, just a thought. If we were to create an installer for JW Lua itself (which would also install the JW Lua script installer), would you @rpatters1 be willing and able to sign it for macOS? I think you have an Apple Developer ID to sign your plugins (let me know if I'm wrong).

If that's possible, it would be extremely helpful. Then we can [create a package using this software or similar](https://www.macupdate.com/app/mac/34613/packages), bypassing all the confusing steps of installing a 3rd-party plugin _and_ a JW Lua script. Such a package could:

1. Detect if JW Lua is already installed
2. If not, install the correct version
3. Install the installer Lua script and add the library path to the `jwpluginstorage.xml` file
4. Prompt users to restart Finale

Basically making the entire process one step for users.

Plus, if it was possible to sign it, we might also be able to bypass the "unverified developer" warning with JW Lua.

Thoughts?

**Comment by rpatters1** [comment]
I am extremely reluctant to sign with my personal apple id. I think it may
go against their terms of use. And anyway, JW Lua ultimately needs to be
signed by Jari if he wants it to continue having a life.

Since all we are trying to do is run some scripts, there has to be a way to
crack this nut without code signing.

**Comment by Nick-Mazuk** [comment]
That's perfectly ok. And yeah, Jari does need to eventually sign JW Lua anyways.

**Comment by Nick-Mazuk** [comment]
Ok, I believe I finally got the file this file to have a list of all the Lua scripts in the repo. It should also update every time we merge into the `master` branch.

https://raw.githubusercontent.com/Nick-Mazuk/jw-lua-scripts/gh-pages/scripts.txt

Here's a sample of the file:

```
FILE=UI_switch_to_selected_part.lua
AUTHOR=CJ Garcia
VERSION=1.2
DATE=June 12, 2020
CATEGORYTAGS=UI
NAME=Switch To Selected Part
DESCRIPTION=Switches to the first part of the top staff in a selected region in a score. Switches back to the score if viewing a part.

FILE=beam_selected_region.lua
AUTHOR=Robert Patterson
VERSION=1.0.1
DATE=June 12, 2020
CATEGORYTAGS=Note
NAME=Beam Selected Region
DESCRIPTION=Beam Selected Region

FILE=beam_unbeam_selected_region.lua
AUTHOR=Robert Patterson
VERSION=1.0.1
DATE=June 12, 2020
CATEGORYTAGS=Note
NAME=Unbeam Selected Region
DESCRIPTION=Unbeam Selected Region
```

This way, we can quickly search for new scripts, check for updates, or anything else. The metadata is taken straight from the plugin def, so anytime we update that, this file will change.

_P.S. If you've gotten notifications about any builds failing, that was me testing Github actions to make sure this always works._

**Comment by rpatters1** [comment]
I was just thinking we should probably require descriptions so people know
why anyone took the trouble to write them.

**Comment by rpatters1** [comment]
P.S. Unfortunately, JW Lua truncates the descriptions at something like 64
characters when it displays them in Finale. That's an ancient limitation
from PDK of ~2003.

**Comment by Nick-Mazuk** [comment]
Created a pull request that updates the README to require descriptions.

Also, tonight I'll go ahead and update all the library import statements to match what we agreed on above.

**Comment by Nick-Mazuk** [comment]
Now that the scripts are bundled to include the applicable functions from the library and users can easily download scripts from the website, shall I close this issue?

Though, a dedicated installer would be nice, I think the current solution is probably good enough. Any thoughts?

**Comment by rpatters1** [comment]
I'm cool with closing it. But maybe CJ should comment as well.

**Comment by Nick-Mazuk** [comment]
Just a note (off topic but this is the best thread):

The `plugindef` can include an option called [`notes`](http://jwmusic.nu/jwplugins/wiki/doku.php?id=jwlua:finaleplugin_properties#notes_multiline_string). This field accepts a multiline string and is correctly parsed and displayed on the website. Might be a good idea to start using this field to document how to use each script. There's only so much you can write in the description of the return statement, and often that's not enough to communicate the full usage of a script.

**Comment by Nick-Mazuk** [comment]
Closing since @CJGarciaMusic hasn't commented on this in a year.

CJ, feel free to reopen.
